util/base/src/main/java/jetbrains/jetpad/base/ThreadSafeAsync.java

```diff
@@ -43,6 +45,20 @@ public Registration onFailure(Handler<Throwable> failureHandler) {
     }
   }
 
+  @Override
+  public <ResultT> Async<ResultT> map(Function<? super ItemT, ? extends ResultT> success) {
+    synchronized (myAsync) {
```

Comment:
Unfortunately, things aren't such easy here. The main problem with thread safe asyncs is that callbacks can be added in one thread, but success() can be called from another, so we needs synchronization on callback list access inside async instance. Asyncs.map() and select() methods create SimpleAsync instances, which have no such synchronization